File: src/main/java/src/MessageBox.java
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 * EXAME PC 2021V_1
 */

public class MessageBox <T> {

    private final Lock lock = new ReentrantLock();
    private List<Box> msgBox = new LinkedList<>();
    private Condition condition = lock.newCondition();

    class Box {
        T msg;
        long timeout;
        Condition cond;
        public Box(long timeout) {
            this.timeout = timeout;
            cond = condition;
        }
    }

    public int sendToAll(T message) {
        int counter = 0;
        if (msgBox.size() > 0) {
            for (Box b : msgBox) {
                b.msg = message;
                counter++;
            }
            condition.signalAll();
        }
        return counter;
    }

    public Optional<T> waitForMessage(long timeout, TimeUnit unit) throws InterruptedException {
        try {
            lock.lock();
            //FAST PATH
            long remainingTime = unit.toNanos(timeout);
            if (remainingTime <= 0) {
                return Optional.empty();
            }
            Box box = new Box(timeout);
            msgBox.add(box);
            if (box.msg != null) {
                msgBox.remove(box);
                return Optional.of(box.msg);
            }
        } finally {
            lock.unlock();
        }
        // SLOW PATH
        while (true) {
            Box box = null;
            try {
                condition.await();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                box = new Box(timeout);
                msgBox.remove(box);
                throw e;
            }
            long remainingTime = unit.toNanos(timeout);
            if (remainingTime <= 0) {
                return Optional.empty();
            }
            assert false;
            if (box.msg != null) {
                msgBox.remove(box);
                return Optional.of(box.msg);
            }
        }
    }
}

File: src/main/kotlin/MessageBox.kt
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.Condition
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock

class MessageBox<T : Any> {
    private val lock: Lock = ReentrantLock()
    private val msgBox: MutableList<Box?> = LinkedList<Box?>()
    private val condition = lock.newCondition()

    internal inner class Box(var timeout: Long) {
        var msg: T? = null
        var cond: Condition = condition


    }

    fun sendToAll(message: T): Int {
        var counter = 0
        if (msgBox.size > 0) {
            for (b in msgBox) {
                if (b != null) {
                    b.msg = message
                }
                counter++
            }
            condition.signalAll()
        }
        return counter
    }

    @Throws(InterruptedException::class)
    fun waitForMessage(timeout: Long, unit: TimeUnit): Optional<T> {
        try {
            lock.lock()
            //FAST PATH
            val remainingTime = unit.toNanos(timeout)
            if (remainingTime <= 0) {
                return Optional.empty()
            }
            val box: Box = Box(timeout)
            msgBox.add(box)
            if (box.msg != null) {
                msgBox.remove(box)
                return Optional.of(box.msg!!)
            }
        } finally {
            lock.unlock()
        }
        // SLOW PATH
        while (true) {
            var box: Box? = null
            try {
                condition.await()
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
                box = Box(timeout)
                msgBox.remove(box)
                throw e
            }
            val remainingTime = unit.toNanos(timeout)
            if (remainingTime <= 0) {
                return Optional.empty()
            }
            if (box != null) {
                if (box.msg != null) {
                    msgBox.remove(box)
                    return Optional.of(box.msg!!)
                }
            }
        }
    }
}